### PR TITLE
fix(desktop): surface each provider's live models in the model picker

### LINKED ISSUE #10260: Desktop model picker lacks search and does not match CLI model discovery (missing newer models)
**Describe the bug**

The model picker in Goose Desktop only shows a small recommended list and does not provide a search path, so new or newly available models are not discoverable in the UI.  
In contrast, the same provider in the CLI (`goose configure`) can find those models via `Search all models...`.  
This appears to be a Desktop/CLI parity issue in model discovery and selection, not an auth or provider-config issue.

---

**To Reproduce**

1. Open Goose Desktop.
2. Go to **Settings → Models → Configure Providers**.
3. Add/select the Google provider (or any provider with many models).
4. Open the model picker in Desktop and view the model list.
5. Observe that only a short recommended list is shown and there is no search/filter input.
6. In a terminal, run `goose configure`.
7. Select the same provider.
8. Choose **Search all models...** from the CLI model selection flow.
9. Compare results: the CLI shows additional/newer models not shown in Desktop.

---

**Expected behavior**

- Desktop should provide model search/filter capability in the picker.
- Desktop should align with CLI model discovery, and expose full or explicitly “all models” results.
- Recommended lists may remain as a default view, but they must not hide newer/available models from selection.
- Any model discoverable from CLI should be selectable in Desktop.

---

**Screenshots**

- Desktop model picker showing the short old/recommended list with no search.
- CLI `Search all models...` showing a fuller/newer model set.

---

**Please provide the following information**
- **OS & Arch:** [ macOS arm64]
- **Interface:** Desktop UI
- **Version:** [v1.41.0]
- **Extensions enabled:** [e.g. none]
- **Provider & Model:** [e.g. Google – gemini-2.5-flash]

---

**Additional context**
- The issue is reproducible with the same account and provider across CLI and Desktop.
- Related references:
  - https://github.com/aaif-goose/goose/issues/9822
  - https://github.com/aaif-goose/goose/issues/10035
  - https://github.com/aaif-goose/goose/releases (includes #7891)
- Diagnostics zip attached if available.

### LINKED ISSUE #10309: Bug: Unable to configure Claude ACP provider due to missing model options and documentation
**Describe the bug**

The model selection box is empty by default, and none of the following model IDs, as specified in [Claude's documentation](https://platform.claude.com/docs/en/about-claude/models/overview), work: `claude-opus-4-8` (Claude API ID), `anthropic.claude-opus-4-8` (AWS Bedrock ID), or `opus`.

Also, valid model IDs are not listed in [Goose's docs](https://goose-docs.ai/docs/guides/acp-providers/#claude-acp-1) or in the [documentation for the underlying library](https://github.com/agentclientprotocol/claude-agent-acp/tree/main/docs).

The error message shown in the screenshot is too simple and lacks diagnostic information, making it unhelpful for troubleshooting.

---

**To Reproduce**
1. Follow the prerequisites documented for the Claude ACP provider.
2. Go to "Settings".
3. Select "Claude Code" (ACP version) in the first select menu. 
   Image
4. Select "Enter a model not listed" in the second select menu.
    Image
5. Manually input `claude-opus-4-8`, `anthropic.claude-opus-4-8`, or `opus`.
6. See the error. 
    Image

Note: I am authenticated in the Claude CLI, and the Claude CLI provider works fine for me.

---

**Expected behavior**

There should be documentation regarding where to look up valid model IDs, and at least "Opus," "Sonnet," and "Haiku" should be listed in the model selection menu.

Additionally, the error message should be descriptive—for example, it should specify if it is a "claude binary not found" error, an "ACP server timeout," or a "Claude API returns 400." If there is a problem with the model field, the application should clearly communicate what the valid values are.

---

**Screenshots**

(Included in steps to reproduce.)

---

**Please provide the following information**
- **OS & Arch:** Windows 11
- **Interface:** UI
- **Version:** 1.40.0
- **Extensions enabled:** Analyze, Apps, Code Mode, Developer, Extension Manager, Memory, Skills, Summon, Todo, Top Of Mind
- **Provider & Model:** Claude Code, N/A

---

**Additional context**
N/A

### PR DESCRIPTION
**Pull request by kojiromike**
## Problem

The desktop Settings model picker gets its list from `acpListProviderModels()` → `providersList_unstable` → the server's provider **inventory cache** (`entry.models`). That cache is:

1. **Async** — filled by a background refresh job that spawns the provider, so it's empty/stale until the refresh completes.
2. **Canonical-filtered** — populated via `fetch_recommended_models()`, which runs every model through the bundled `canonical_models.json`.

So newly released models (e.g. `claude-opus-5`) and adapter-advertised aliases from the **Claude Code (ACP)** provider (`opus[1m]`, `sonnet`, `claude-fable-5[1m]`, `default`, …) get silently dropped or never appear — even though the backend can enumerate them.

Meanwhile a purpose-built RPC already exists on both sides and returns the **raw, live, unfiltered** list — `_goose/unstable/providers/supported-models/list` → `on_list_provider_supported_models` → `fetch_supported_models()` — but the desktop never called it.

Fixes #10309 and #10691. Fixes #10260.

## Fix

Wire the picker to that endpoint. `acpListProviderModels` now **unions** the live supported-models list with the inventory cache:

- Live ids lead, in their advertised order (freshest first — e.g. Anthropic's `/v1/models` is now newest-first).
- Cached metadata (context limit, reasoning) is grafted onto matching ids; live-only ids get a synthetic entry.
- Cached-only models are appended, so nothing that showed before disappears (purely additive).
- Live discovery is best-effort — on error it falls back to the inventory cache unchanged.

Applied to **all providers**, so it also surfaces first-party API models absent from the bundled canonical registry (the #8321 class of drops), not just the ACP path.

No Rust changes — the endpoint was already there; this connects the existing wire.

## Testing

- New unit tests in `providers.test.ts` cover the union ordering/metadata-graft and the fallback-on-error path.
- `pnpm typecheck`, `eslint`, and `vitest` all pass.

## Notes / follow-ups

- This makes models **selectable/visible**; a live-only model's context limit still falls back to the default until the canonical registry knows it (the `max_input_tokens` enrichment is a separate follow-up).
- Complements #10675 (which makes the Claude Code ACP picker's *selection* actually reach the adapter) — together: the list is complete **and** the selection sticks.

### CODE REVIEW

File: ui/desktop/src/acp/providers.ts [hunk 1]
```diff
@@ -62,10 +63,48 @@ export async function acpListProviderDetails(): Promise<ProviderDetails[]> {
...
+  const inventory = entries.find((e) => e.providerId === providerId)?.models ?? [];
+
+  // The inventory cache is canonical-filtered and populated by a background refresh, so it
+  // lags new releases and can silently drop models the key/adapter actually has (e.g. Opus 5,
+  // or an ACP adapter's advertised aliases). Union in the live supported-models list so the
+  // picker never hides a currently-available model. Live ids lead (freshest ordering); cached
+  // metadata (context limit, reasoning) is grafted on by id where we have it.
+  let liveIds: string[];
+  try {
+    liveIds = await acpListProviderSupportedModels(providerId);
```
[the remark this answers is not in the document]
Comment (kojiromike) [reply ui/desktop/src/acp/providers.ts]: Fixed in e4a1d88. Live discovery is now wrapped in a 4s `withTimeout`; on timeout (or error) we fall back to the cached inventory, so a provider that accepts the request but never completes model discovery (e.g. Claude Code's indefinitely-waiting control request) can no longer leave the picker's `Promise.all` loading forever. Added a fake-timer test that asserts fallback-to-inventory when live discovery never resolves.

File: ui/desktop/src/acp/providers.ts [hunk 2]
```diff
@@ -62,10 +63,48 @@ export async function acpListProviderDetails(): Promise<ProviderDetails[]> {
...
+    // Live discovery is best-effort; fall back to the inventory cache when it's unavailable.
+    return inventory;
+  }
+
+  const inventoryById = new Map(inventory.map((model) => [model.id, model]));
+  const seen = new Set<string>();
+  const merged: ProviderInventoryModelDto[] = [];
+  for (const id of liveIds) {
+    seen.add(id);
+    merged.push(inventoryById.get(id) ?? { id, name: id });
```
[the remark this answers is not in the document]
Comment (kojiromike) [reply ui/desktop/src/acp/providers.ts:156]: Reworded in e4a1d88 to stop over-claiming. The merge preserves whatever order `liveIds` arrives in; per-provider recency ordering is the provider's responsibility, not this desktop code's. For the native Anthropic provider that ordering is fixed separately in #10722, which replaces the alphabetical `models.sort()` you cite with a newest-first sort on `created_at`. The comment now reads "live ids lead, in the order the backend returns them — each provider owns its own ordering".

**Comment by kojiromike** [comment]
Addressed the timeout and ordering feedback in e4a1d88. @codex review

File: ui/desktop/src/acp/providers.ts [hunk 3]
```diff
@@ -62,10 +63,76 @@ export async function acpListProviderDetails(): Promise<ProviderDetails[]> {
...
+    // Live discovery is best-effort; fall back to the inventory cache when it's slow or fails.
+    return inventory;
+  }
+
+  const inventoryById = new Map(inventory.map((model) => [model.id, model]));
+  const seen = new Set<string>();
+  const merged: ProviderInventoryModelDto[] = [];
+  for (const id of liveIds) {
+    seen.add(id);
+    merged.push(inventoryById.get(id) ?? { id, name: id });
```
[the remark this answers is not in the document]
Comment (kojiromike) [reply ui/desktop/src/acp/providers.ts:156]: Good catch — narrowed the union rather than filtering raw IDs. `acpListProviderModels` now unions the live list only for providers whose supported-models endpoint returns exclusively agent-usable models: ACP agent adapters (`category === 'agent'`, generic across current/future adapters) and the first-party Anthropic API (allowlisted; its `/v1/models` is all chat/tool-capable). Other model providers (OpenAI, Google, OpenRouter, …) keep the canonical-filtered inventory list, so embeddings/audio/image models no longer leak into the picker. Added tests for both the agent-category union and the non-allowlisted no-union case (which also asserts live discovery is skipped entirely).

File: ui/desktop/src/acp/providers.ts [hunk 4]
```diff
@@ -62,10 +63,76 @@ export async function acpListProviderDetails(): Promise<ProviderDetails[]> {
...
+  // returns them — each provider owns its own ordering (the Anthropic provider sorts
+  // newest-first); cached metadata (context limit, reasoning) is grafted on by id where we
+  // have it.
+  let liveIds: string[];
+  try {
+    liveIds = await withTimeout(
+      acpListProviderSupportedModels(providerId),
+      LIVE_MODELS_TIMEOUT_MS,
+      `live model discovery for ${providerId}`
+    );
```
[the remark this answers is not in the document]
Comment (kojiromike) [reply ui/desktop/src/acp/providers.ts:145]: Right — the desktop 4s timeout only abandoned the client promise. Added a server-side bound in `on_list_provider_supported_models`: `tokio::time::timeout(SUPPORTED_MODELS_TIMEOUT /* 10s */, provider.fetch_supported_models())`. On elapse the future is dropped, which tears down the outstanding fetch and (for Claude Code) its spawned child via the provider's `Drop`, instead of leaking a permanent request/subprocess. The desktop timeout stays as defense-in-depth.

**Comment by kojiromike** [comment]
@codex review

File: crates/goose/src/acp/server/providers.rs [hunk 5]
```diff
@@ -478,10 +485,17 @@ impl GooseAcpAgent {
             .create_provider(&req.provider_id, Vec::new(), None)
             .await
             .internal_err_ctx("Failed to initialize provider")?;
-        let models = provider
-            .fetch_supported_models()
-            .await
-            .internal_err_ctx("Failed to fetch provider supported models")?;
+        let models =
+            tokio::time::timeout(SUPPORTED_MODELS_TIMEOUT, provider.fetch_supported_models())
+                .await
```
[the remark this answers is not in the document]
Comment (kojiromike) [reply crates/goose/src/acp/server/providers.rs]: Fixed in 7249a35. `create_provider()` is now inside the timeout — both it and `fetch_supported_models()` run in a single `async` block wrapped by `tokio::time::timeout(SUPPORTED_MODELS_TIMEOUT, …)`, so an adapter that hangs during initialization or `session/new` is bounded by the same 10s deadline rather than never starting the clock.

[the remark this answers is not in the document]
Comment (kojiromike) [reply crates/goose/src/acp/server/providers.rs]: Right — the explicit `child.kill().await` in `fetch_supported_models` only runs if `exchange_control` returns, and a timed-out caller drops the future before that. Fixed in 7249a35 by setting `cmd.kill_on_drop(true)` on the model-listing command before spawn, so dropping the handle reaps the Claude process instead of orphaning it. Scoped to the model-list child rather than `build_stream_json_command()` itself, to leave the long-lived `spawn_process` child's lifecycle unchanged.

**Comment by kojiromike** [comment]
Addressed both P2 findings in 7249a35 — server timeout now covers provider startup, and the model-list child is `kill_on_drop`. @codex review

File: ui/desktop/src/acp/providers.ts [hunk 6]
```diff
@@ -62,10 +64,101 @@ export async function acpListProviderDetails(): Promise<ProviderDetails[]> {
...
+// their request timeout. Bound it so a single slow provider can't leave the whole picker
+// (RecipeModelSelector awaits every provider via Promise.all) loading forever.
+const LIVE_MODELS_TIMEOUT_MS = 4000;
+
+// Model providers whose live supported-models endpoint returns *only* agent-usable models.
+// The first-party Anthropic API's /v1/models is all chat/tool-capable, so its raw list is safe
+// to surface directly. Other model providers (OpenAI, Google, OpenRouter, …) also return
+// embeddings, audio, and image models from their raw endpoint, which would fail an agent
+// request; for those we keep the canonical-filtered inventory list instead.
+const LIVE_UNION_MODEL_PROVIDERS = new Set(['anthropic']);
```
[the remark this answers is not in the document]
Comment (kojiromike) [reply ui/desktop/src/acp/providers.ts]: Confirmed and fixed in 10df531. `claude-code` is still registered (`init.rs:84`), and `get_provider_setup_category` matches on exact `provider_id` only — the catalog's `claude_code` alias on the `claude-acp` entry doesn't match the hyphenated id, so it falls through to `ProviderSetupCategory::Model` (`inventory/mod.rs:719-720`) and never reached the live path. Added `'claude-code'` to `LIVE_UNION_MODEL_PROVIDERS` with a comment explaining why an agent provider needs the model-provider allowlist, plus a test covering the empty-inventory + live-aliases case.

File: crates/goose/src/acp/server/providers.rs [hunk 7]
```diff
@@ -474,14 +481,27 @@ impl GooseAcpAgent {
...
+            let provider = self
+                .create_provider(&req.provider_id, Vec::new(), None)
+                .await
+                .internal_err_ctx("Failed to initialize provider")?;
+            provider
+                .fetch_supported_models()
+                .await
+                .internal_err_ctx("Failed to fetch provider supported models")
+        };
+        let models = tokio::time::timeout(SUPPORTED_MODELS_TIMEOUT, discover)
```
[the remark this answers is not in the document]
Comment (kojiromike) [reply crates/goose/src/acp/server/providers.rs:496]: You're right — `spawn_client_loop` uses `std::thread::spawn` with its own current-thread runtime, so dropping the `start()` future can't abort it, and `handle_requests` awaited `initialize` (and `session/new`) with no bound and no select on `rx` closure. The `Child` is held by `run_with_child` on that thread, so a wedged handshake pinned the adapter process indefinitely.

Fixed in 10df531 by putting the cancellation where the child actually lives rather than at the caller: both startup requests are now wrapped in `tokio::time::timeout(STARTUP_TIMEOUT /* 60s */, …)`. On elapse `initialize` returns an error out of `connect_with`, which unwinds to `run_with_child`'s `child.kill().await` / `child.wait().await` and ends the thread; `session/new` reports the timeout to its `response_tx` so the loop stays responsive and unwinds when `start()` drops `tx`. `spawn_acp_process` already sets `kill_on_drop(true)`, so the child is reaped either way. 60s is a wedge backstop for a cold adapter start, not a latency budget — the 10s RPC bound and 4s desktop bound still govern picker responsiveness.

**Comment by kojiromike** [comment]
Addressed both P2 findings in 10df531 — `claude-code` added to the live-model allowlist, and the ACP startup handshake (`initialize`, `session/new`) is now bounded inside the client loop so a wedged adapter reaps its own child. @codex review

File: crates/goose/src/acp/provider.rs [hunk 8]
```diff
@@ -1093,23 +1116,30 @@ async fn handle_requests(
...
+                    .block_task(),
+                )
+                .await;
                 let result = match session {
-                    Ok(session) => {
+                    Ok(Ok(session)) => {
                         session_ids.push(session.session_id.clone());
                         apply_session_config_options(&config, &cx, session.session_id.clone())
                             .await?;
                         apply_session_mode(&config, &goose_mode, &cx, session).await
```
[the remark this answers is not in the document]
Comment (kojiromike) [reply crates/goose/src/acp/provider.rs]: Correct — I bounded `session/new` itself but left `apply_session_config_options` and `apply_session_mode` outside the deadline, so an adapter that answered `session/new` and then wedged on `session/set-config-option` or `session/set-mode` still pinned its child.

Fixed in ead0dd0. The whole startup exchange now runs inside a single `tokio::time::timeout(STARTUP_TIMEOUT, …)` — `session/new`, the config options, and the mode application — rather than one timeout per request, so the bound is on total startup rather than per-hop. `session_ids.push` stays immediately after `session/new` so a session that fails during setup is still tracked for close.

One behavior change worth calling out: `apply_session_config_options(...).await?` previously propagated out of `handle_requests` and tore down the client loop. It's now reported through `response_tx` like the mode path. Net effect is the same — `start()` gets the error and drops `tx`, which unwinds the loop — but the failure is now attributable to the caller instead of surfacing as a bare loop exit.

`cargo clippy -p goose --all-targets` clean; `acp::` lib tests (197) and the `acp_provider_test` integration suite, which drives this exact startup path over a real transport, both pass.

**Comment by kojiromike** [comment]
Addressed in ead0dd0 — the startup bound now covers the post-`session/new` setup RPCs (config options, mode) under a single deadline, not just `session/new`. @codex review

**Comment by DOsinga** [comment]
Thanks for digging into live provider models and the desktop picker. The original issue was closed in favor of #11030, which better captures the problem and is now the place to agree on scope and verification. I’m closing this PR so future work can follow that issue. We appreciate the investigation.
